### PR TITLE
feat(library/algebra/group_power): add times (additive variant of power)

### PR DESCRIPTION
**Pull request by fgdorais**
Following common usage, I made `times` act on the left rather than acting on the right, like `pow`. I adjusted the order of arguments to all theorems to match.

### CODE REVIEW

**Comment by avigad** [comment]
@fgdorais This is a nice addition. Thanks!

As I was driving in two work this morning, I thought about it, and two things occurred to me. The first is minor, the second is more important.

First, I think I would prefer to use `smul` for scalar multiplication instead of `times`. It seems awkward to use both `times` and `mul` to describe different kinds of multiplication, and also one would expect `times` to have two arguments of the same type. The phrase "scalar multiplication" is familiar and more descriptive.

Second, I am worried about further overloading `*` whenever this file is imported and algebra is open. At times, it will even be ambiguous: another way to interpret `3 * x` in a ring is to coerce `3` to an element of the ring and carry out the usual multiplication there. One solution is to put the notation in a namespace so that the user has to open it explicitly. I think a better solution is to use another symbol, like `⬝`. (This is currently entered as \cdot, but we can give it other names as well.)

In short, I am proposing to
- rename `times` to `smul` (and revise the theorem names accordingly)
- replace `*` by `⬝`
  Please let me know if you see any problems with this.

By the way, I am also planning to rename the file `monoid_power`, and move the group power operation to theory/finite_groups. Using negative exponents in that way is more specialized to group theory. We should also define `field_power`, so that we can use negative exponents with the rationals and reals. I will keep that in the algebra folder.

**Comment by fgdorais** [comment]
@avigad Happy to help!

I'm happy with `smul` and your rationale makes a lot of sense, though I'm less happy with `ismul` which suggests a proposition `is_mul`. Another option is to make the left type explicit, for example: `ntimes`/`itimes`, `nat_times`/`int_times`, `nmul`/`imul`.

I was worried about overloading `*` too and I was surprised that there were no issues. I'm happy with using `⬝` instead and I think that will also improve overall readability.

I'm less sure about your renaming plan. It seems to me that algebra is a fine place for this since it has broad applicability. It seems odd to have to import theory/finite_groups to get `ipow`.

**Comment by avigad** [comment]
Ah, I forgot about the "i" variant. I'd like to keep the names "pow" and "smul" for the nat versions, because they are the most commonly used.

We will have a similar problem when we define integer power for fields, which will be a different notion, since we will want to define a value for 0. How about `gpow` and `gsmul` for integer power and scalar multiplication in a group, and `fpow` for integer power in a field? (Fortunately or unfortunately, gsmul will work on the additive part of a field, so users will have to get used to that. Anyhow, in classes like the reals etc. I think it will be more common to write n \* a and have n coerced to a real.)

You have convinced me that it is worthwhile keeping integer powers in algebra, so I'll leave them there, whatever we end up calling them.

**Comment by fgdorais** [comment]
@avigad I think `gpow` and `fpow` make sense as a way of distinguishing the group and field cases. What do Coq and Isabelle do? Perhaps it's worth thinking about what people who will write algebra/ring_module or theories/modules would prefer? If that were me, I would want to use `smul` for scalar multiplication on modules and I might find it annoying that `smul` is already defined.

Let me make a quick visual argument in favor of `ntimes`:
t

At this point, I've thought about this a little too much to think any option is much better than the others. Your global vision of the lean library is much better than mine and I'm happy to go with what you think is best.

**Comment by avigad** [comment]
I see, people will want to use smul for scalar multiplication with the associated ring, not just scalar multiplication by natural numbers. Good point. Life is so complicated!

I am now inclined to go with the following scheme:

  nmul : scalar multiplication by a natural number
  imul  : scalar multiplication by an integer
  smul : general scalar multiplication

as well as

  pow   : natural number power in any monoid
  gpow : integer power in a group
  fpow  : integer power in a field

It's not perfect but it does group the analogous concepts together and the mnemonics make sense. I think it will work.

**Comment by fgdorais** [comment]
@avigad I don't see any immediate issues with this scheme. I'm on board!
